src/aatoolbox/utils/io.py

```diff
@@ -82,35 +83,48 @@ def parse_yaml(filename: Union[str, Path]) -> dict:
     return config
 
 
-def check_file_existence(func: F) -> F:
+@wrapt.decorator
+def check_file_existence(wrapped: F, instance, args, kwargs) -> F:
```

Comment:
Is it possible to type these arguments somehow? Like should args be a list and kwargs a dict? And could instance be `DataSrouce` (would it then accept child classes?)

Reply:
I am trying to figure out, what IS self for passing? I think we are mainly concerned with decorating instance methods, right, rather than class methods. How do you type hint self? It isn't a class I think, right? Like somehow it's different? Or if it's self do we just try `Type[DataSource]` to specify that it's [a subclass of DataSource].(https://stackoverflow.com/questions/46092104/subclass-in-type-hinting)

Reply:
Figured out the typing, I think we can directly pass the abstract class and require it. I think it is working (according to tests). See the thread [here](https://stackoverflow.com/questions/38791739/function-annotation-for-subclasses-of-abstract-class).